README.md
# MagicMirror Module: MMM-YouTubeChannelStats

A MagicMirror Module for displaying statistics about YouTube channels.

[![Platform](https://img.shields.io/badge/platform-MagicMirror-informational)](https://MagicMirror.builders)
[![License](https://img.shields.io/badge/license-MIT-informational)](https://raw.githubusercontent.com/ianperrin/MMM-YouTubeChannelStats/master/LICENSE)
![Test Status](https://github.com/ianperrin/MMM-YouTubeChannelStats/actions/workflows/node.js.yml/badge.svg)
[![Code Climate](https://codeclimate.com/github/ianperrin/MMM-YouTubeChannelStats/badges/gpa.svg)](https://codeclimate.com/github/ianperrin/MMM-YouTubeChannelStats)
[![Known Vulnerabilities](https://snyk.io/test/github/ianperrin/MMM-YouTubeChannelStats/badge.svg)](https://snyk.io/test/github/ianperrin/MMM-YouTubeChannelStats)

## Example

![Example screenshot](.github/example.gif)

## Installation

1. Stop your MagicMirror and clone the repository into the modules folder

   ```bash
   cd ~/MagicMirror/modules
   git clone https://github.com/ianperrin/MMM-YouTubeChannelStats.git
   cd ~/MagicMirror/modules/MMM-YouTubeChannelStats
   ```

2. Create a Google API key

   - For more information see the Google documentation for [obtaining a developer key](https://developers.google.com/youtube/v3/getting-started)
   - Make a note of the `API key`

3. Add the module to the config file (`~/MagicMirror/config/config.js`) for your mirror.

   ```javascript
   modules: [
     {
       module: "MMM-YouTubeChannelStats",
       position: "top_right",
       config: {
         channelIds: "your_youtube_channel_id,another_youtube_channel_id",
         apiKey: "your_google_api_key"
       }
     }
   ];
   ```

   The full list of config options can be found in the [configuration options](#configuration-options) table.

4. Restart the MagicMirror

   ```bash
   pm2 restart mm
   ```

## Updating the module

To update the module to the latest version,

1. Pull the changes from this repository into the module folder:

   ```bash
   cd ~/MagicMirror/modules/MMM-YouTubeChannelStats
   git pull
   ```

2. Restart the mirror

   ```bash
   pm2 restart mm
   ```

## Configuration options

The following properties can be added to the configuration:

| **Option**         | **Default**                          | **Description**                                                                                                                                                                                                                                                           | **Possible Values**                                   |
| ------------------ | ------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------- |
| `channelIds`       |                                      | The YouTube channel id or ids. _Note:_ Multiple channel ids can be entered as a comma separated list (e.g. `"id1,id2"`) or an array of strings (e.g. `["id1", "id2"]`)..                                                                                                  |                                                       |
| `apiKey`           |                                      | The YouTube API key.                                                                                                                                                                                                                                                      |                                                       |
| `stats`            | `["views", "subscribers", "videos"]` | _Optional_ - Which statistics to display. _Note:_ - The stats can be listed in any order, and only one is required. However, they must be entered as an array of strings i.e. comma separated values within square brackets.                                              | `views`, `subscribers`, `videos`                      |
| `showLabels`       | `true`                               | _Optional_ - Display labels next to the statistics.                                                                                                                                                                                                                       | `true` or `false`                                     |
| `showChannelThumbnail`       | `true`                     | _Optional_ - Display channel thumbnail next to the statistics.                                                                                                                                                                                                                       | `true` or `false`                                     |
| `grayscale`        | `true`                               | _Optional_ - Display channel images in grayscale.                                                                                                                                                                                                                         | `true` or `false`                                     |
| `maximumChannels`  | `0`                                  | _Optional_ - The maximum number of channels displayed in the module. If `0` (or not defined), all channels will be shown. If `maximumChannels` is less than the number of configured `channelIds`, the channels will automatically rotate based on the `rotationInterval` | `0` or a number less than the number of `channelIds`. |
| `fetchInterval`    | `3600 * 1000` (1 hour)               | _Optional_ - How often the statistics should be fetched (in milliseconds).                                                                                                                                                                                                | `1000` - `86400000`                                   |
| `rotationInterval` | `10 * 1000` (10 seconds)             | _Optional_ - How often the channels should be rotated if `maximumChannels` is also set.                                                                                                                                                                                   | `1000` - `86400000`                                   |
| `animationSpeed`   | `2.5 * 1000` (2.5 seconds)           | _Optional_ - The speed of the update animation (in milliseconds).                                                                                                                                                                                                         | `0` - `5000`                                          |


CHANGELOG.md
# MMM-YouTubeChannelStats Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](https://semver.org/).

## [1.4.0] - Unreleased (Develop Branch)

### Added

- Code coverage support (`npm run test:coverage`)
- Unit test coverage for inherited functions in the core module file.
- Added `showChannelThumbnail` config option.

### Updated

- `getOrMakeArray` function now returns an empty array when passed a zero length string

### Removed

### Fixed

## [1.3.0] - 2021-04-05

### Added

- The number of channels displayed can be limited using the `maxChannels` configuration option.
- The channels displayed will automatically rotated if the number of `channelIds` exceeds `maxChannels`.
- The `grayscale` configuration option can be used to control the appearance of the channels thumbnail image. By default, thumbnails will be shown in grayscale. When set to `false`, thumbnails will appear as on YouTube.
- Added an animated example screenshot to illustrate multiple channel support, channel rotation and thumbnail position/display.
- Unit tests for custom functions in the core module file.

### Changed

- Channel `stats` are now formatted to reflect how they [appear on YouTube](https://developers.google.com/youtube/v3/revision_history#september-10,-2019).
- Multiple `channelIds` can now be supplied as a string (in CSV format) as well as a string array.
- Refactored the functions used to call the API.

## [1.2.0] - 2021-04-03

### Added

- Multiple YouTube channels can be displayed within one instance of the module using the new `channelIds` configuration option.

### Changed

- The channel thumbnail now appears to the right of the statistics when the module is positioned on the right of the MagicMirror.

### Removed

- The `comments` value is no longer supported in the `stats` configuration option as `subscriber.commentCount` has been [deprecated by the API](https://developers.google.com/youtube/v3/revision_history#september-9,-2020).
- Removed the `channelId` configuration option. Existing implementations should be updated to use the new `channelIds` configuration option instead.
- Removed the `setHeader` configuration option. The YouTube channel title is now added as a header within the module above each channel.

## [1.1.0] - 2021-04-02

### Added

- Added Github action for automated linting using prettier, eslint and stylelint.
- Added French, German, Italian and Spanish translations.

## [1.0.0] - 2020-07-26

### Initial release of the MMM-YouTubeChannelStats for MagicMirror


tests/MMM-YouTubeChannelStats.test.js
const chai = require("chai");
const expect = require("chai").expect;
const chaiAsPromised = require("chai-as-promised");
const sinon = require("sinon");
const moduleAlias = require("module-alias");
moduleAlias.addAliases({ logger: "../../../js/logger.js" });
global.Log = require("logger");
global.fetch = require("node-fetch");
const moduleName = "MMM-YouTubeChannelStats";

describe(`Functions in ${moduleName}.js`, function () {
	Module = {};
	const channelIds = {
		channelid1: { return: ["channelid1"] },
		"channelid1,channelid2": { return: ["channelid1", "channelid2"] },
		"channelid1, channelid2": { return: ["channelid1", "channelid2"] }
	};
	before(function () {
		Module.definitions = {};
		Module.register = function (name, moduleDefinition) {
			Module.definitions[name] = moduleDefinition;
		};
		require(`../${moduleName}.js`);
		Module.definitions[moduleName].name = moduleName;
		Module.definitions[moduleName].identifier = "module_1_" + moduleName;
		Module.definitions[moduleName].nunjucksEnvironment = function () {
			return { addFilter: function () {} };
		};
		Module.definitions[moduleName].updateDom = function () {};
		Module.clock = sinon.useFakeTimers(Date.now());
		chai.use(chaiAsPromised);
	});
	after(function () {
		Module.clock = sinon.restore();
	});
	describe("start", function () {
		let moduleConfig = {};
		before(function () {
			Module.definitions[moduleName].config = Module.definitions[moduleName].defaults;
		});
		describe("for default config", function () {
			before(function () {
				Module.definitions[moduleName].start();
				moduleConfig = Module.definitions[moduleName].config;
				Module.clock.tick(moduleConfig.fetchInterval + 1);
			});
			it("channelIds should be an array with a length of 0", function () {
				expect(moduleConfig.channelIds).to.be.an("array").with.a.lengthOf(0);
			});
			it("stats should be an array with a length of 2", function () {
				expect(moduleConfig.stats).to.be.an("array").with.a.lengthOf(3);
			});
			it("enableRotate should be false", function () {
				expect(moduleConfig.enableRotate).to.be.false;
			});
		});
		describe("for custom config", function () {
			before(function () {
				Module.definitions[moduleName].config.channelIds = Object.keys(channelIds)[1];
				Module.definitions[moduleName].config.maximumChannels = 1;
				Module.definitions[moduleName].start();
				moduleConfig = Module.definitions[moduleName].config;
				Module.clock.tick(moduleConfig.fetchInterval);
			});
			it("channelIds should be an array with a length of 2", function () {
				expect(moduleConfig.channelIds).to.be.an("array").that.has.a.lengthOf(2);
			});
			it("stats should be an array with a length of 2", function () {
				expect(Module.definitions[moduleName].config.stats).to.be.an("array").that.has.a.lengthOf(3);
			});
			it("enableRotate should be true", function () {
				expect(moduleConfig.enableRotate).to.be.true;
			});
		});
	});
	describe("getStyles", function () {
		it("to register font-awesome and custom css", function () {
			expect(Module.definitions[moduleName].getStyles()).to.eql(["font-awesome.css", "MMM-YouTubeChannelStats.css"]);
		});
	});
	describe("getTemplate", function () {
		it("to return custom njk template", function () {
			expect(Module.definitions[moduleName].getTemplate()).to.equal("templates\\statistics.njk");
		});
	});
	describe("getTemplateData", function () {
		let templateData = {};
		before(function () {
			Module.definitions[moduleName].data = { position: "top_right" };
			templateData = Module.definitions[moduleName].getTemplateData();
		});
		it("for object to have expected keys", function () {
			expect(templateData).to.have.keys(["channelsList", "config", "position"]);
		});
		it("for 'position' to equal 'right'", function () {
			expect(templateData.position).to.equal("right");
		});
		it("for 'config' to equal module config", function () {
			expect(templateData.config).to.eql(Module.definitions[moduleName].config);
		});
	});
	describe("getTranslations", function () {
		let translations = {};
		let translationFiles = [];
		before(function () {
			const { readdirSync } = require("fs");
			const { join } = require("path");
			translationFiles = readdirSync(join(__dirname, "..", "translations"));
			translations = Module.definitions[moduleName].getTranslations();
		});

		it(`to have translation keys with matching translation file`, function () {
			const expectedKeysForTranslationFiles = translationFiles.map((translationFile) => translationFile.split(".")[0]);
			expect(translations).to.have.keys(expectedKeysForTranslationFiles);
		});
		it(`to have translation files for each translation key`, async function () {
			const expectedFilesForTranslationKeys = Object.keys(translations).map((translationKey) => translations[translationKey].split("/")[1]);
			expect(translationFiles).to.have.members(expectedFilesForTranslationKeys);
		});
	});
	describe("fetchData", function () {
		it(`for missing 'apiKey' to be undefined`, function () {
			Module.definitions[moduleName].config.apiKey = "";
			expect(Module.definitions[moduleName].fetchData()).to.be.undefined;
		});
		it(`for invalid 'apiKey' to be undefined`, function () {
			Module.definitions[moduleName].config.apiKey = "blah";
			//			expect(Module.definitions[moduleName].fetchData()).to.eventually.be.undefined;
		});
	});
	describe("getParams", function () {
		Object.keys(channelIds).forEach((channelId) => {
			it(`for '${channelId}' params contains '${channelId.replace(" ", "")}'`, function () {
				Module.definitions[moduleName].config.channelIds = channelIds[channelId].return;
				expect(Module.definitions[moduleName].getParams()).contains("?id=" + channelId.replace(" ", ""));
			});
		});
	});
	describe("processData", function () {
		it(`for 'valid responses' should set 'channelsList'`, function () {
			const validResponse = {
				status: 200,
				json: function () {
					return { items: ["channel1", "channel2"] };
				}
			};
			Module.definitions[moduleName].config.enableRotate = false;
			Module.definitions[moduleName].processData(validResponse.json());
			expect(Module.definitions[moduleName].channelsList).to.eql(validResponse.json().items);
		});
		it(`for 'valid responses' with rotation should set 'channelsList' and 'currentChannelList'`, function () {
			const validResponse = {
				status: 200,
				json: function () {
					return { items: ["channel1", "channel2"] };
				}
			};
			Module.definitions[moduleName].config.enableRotate = true;
			Module.definitions[moduleName].config.maximumChannels = 1;
			Module.definitions[moduleName].processData(validResponse.json());
			Module.clock.tick(Module.definitions[moduleName].config.rotateInterval);
			expect(Module.definitions[moduleName].channelsList).to.eql(validResponse.json().items);
			expect(Module.definitions[moduleName].currentChannelList.items).to.eql(validResponse.json().items.slice(1));
		});
		it(`for 'badRequest' should throw error`, function () {
			const badRequest = {
				status: 403,
				json: function () {
					return { error: { message: "badRequest" } };
				}
			};
			expect(() => Module.definitions[moduleName].processData(badRequest.json())).to.throw("Unable to process data: badRequest");
		});
	});
	describe("paginateData", function () {
		const arrays = [["ch1"], ["ch1", "ch2"], ["ch1", "ch2", "ch3"], ["ch1", "ch2", "ch3", "ch4"], ["ch1", "ch2", "ch3", "ch4", "ch5"], ["ch1", "ch2", "ch3", "ch4", "ch5", "ch6"], ["ch1", "ch2", "ch3", "ch4", "ch5", "ch6", "ch7"]];
		arrays.forEach((array) => {
			describe(`for [${array}]`, function () {
				[1, 2, 3, 4, 5, 6].forEach((perPage) => {
					it(`using ${perPage} per page, page 1 should return [${array.slice(0, perPage)}]`, function () {
						expect(Module.definitions[moduleName].paginateData(array, 1, perPage).items).to.eql(array.slice(0, perPage));
					});
				});
			});
		});
	});
	describe("getOrMakeArray", function () {
		describe("when the input is a string", function () {
			it(`for '' should return an array with a length of 0`, function () {
				expect(Module.definitions[moduleName].getOrMakeArray("")).to.be.an("array").with.a.lengthOf(0);
			});
			Object.keys(channelIds).forEach((channelId) => {
				it(`for '${channelId}' should return [${channelIds[channelId].return}]`, function () {
					expect(Module.definitions[moduleName].getOrMakeArray(channelId)).to.eql(channelIds[channelId].return);
				});
			});
		});
		describe("when the input is an array", function () {
			const arrays = [["channelid1"], ["channelid1", "channelid2"], [1, "channelid2"]];
			arrays.forEach((array) => {
				it(`for [${array}] should return itself`, function () {
					expect(Module.definitions[moduleName].getOrMakeArray(array)).to.eql(array);
				});
			});
		});
	});
	describe("toMetric", function () {
		const numbers = {
			"-12": { return: "-12" },
			0: { return: "0" },
			1: { return: "1" },
			12: { return: "12" },
			123: { return: "123" },
			1234: { return: "1.23K" },
			12345: { return: "12.3K" },
			123456: { return: "123K" },
			1234567: { return: "1.23M" },
			12345678: { return: "12.3M" },
			123456789: { return: "123M" },
			1230: { return: "1.23K" },
			12300: { return: "12.3K" },
			123000: { return: "123K" },
			1230000: { return: "1.23M" },
			12300000: { return: "12.3M" },
			123000000: { return: "123M" }
		};
		Object.keys(numbers).forEach((num) => {
			it(`for '${num}' should return [${numbers[num].return}]`, function () {
				expect(Module.definitions[moduleName].toMetric(num)).to.eql(numbers[num].return);
			});
		});
	});
});


MMM-YouTubeChannelStats.js
Module.register("MMM-YouTubeChannelStats", {
	// Default module config.
	defaults: {
		channelIds: "",
		apiKey: "",
		stats: ["views", "subscribers", "videos"], // possible values "views", "comments", "subscribers", "videos"
		showLabels: true,
		showChannelThumbnail: true,
		grayscale: true,
		maximumChannels: 0,
		fetchInterval: 3600 * 1000, // 1 hour
		rotateInterval: 10 * 1000, // 10 seconds
		animationSpeed: 2.5 * 1000, // 2.5 seconds
		apiBase: "https://www.googleapis.com/youtube/",
		apiVersion: "v3",
		apiEndpoint: "channels"
	},
	start: function () {
		Log.info("Starting module: " + this.name);
		// Validate config
		this.config.channelIds = this.getOrMakeArray(this.config.channelIds || this.config.channelId);
		this.config.stats = this.config.stats.map((stat) => stat.toLowerCase());
		// Enable rotation if maximumChannels is set and is greater than the number of channels
		this.config.enableRotate = this.config.maximumChannels > 0 && this.config.maximumChannels < this.config.channelIds.length;
		// Add custom nunjucks filters
		this.addFilters();
		// Schedule the update.
		this.scheduleUpdate();
	},
	getStyles: function () {
		return ["font-awesome.css", "MMM-YouTubeChannelStats.css"];
	},
	getTemplate: function () {
		return "templates\\statistics.njk";
	},
	getTemplateData: function () {
		return {
			channelsList: this.currentChannelList ? this.currentChannelList.items : this.channelsList,
			config: this.config,
			position: this.data.position.includes("_right") ? "right" : "left"
		};
	},
	getTranslations: function () {
		return {
			en: "translations/en.json",
			de: "translations/de.json",
			es: "translations/es.json",
			fr: "translations/fr.json",
			it: "translations/it.json"
		};
	},
	addFilters() {
		var env = this.nunjucksEnvironment();
		env.addFilter("toMetric", this.toMetric.bind(this));
	},
	scheduleUpdate: function () {
		// fetch data from API
		this.fetchData();
		setInterval(() => {
			this.fetchData();
		}, this.config.fetchInterval);
	},
	scheduleRotate: function () {
		// rotate data in module
		if (this.config.enableRotate) {
			this.rotationInterval = setInterval(() => {
				this.rotateData();
			}, this.config.rotateInterval);
		}
	},
	fetchData: function () {
		if (this.config.apiKey === "") {
			Log.error(this.identifier, "apiKey not set.");
			return;
		}
		clearInterval(this.rotationInterval);
		const url = this.config.apiBase + this.config.apiVersion + "/" + this.config.apiEndpoint + this.getParams();
		fetch(url)
			.then((response) => response.json())
			.then((data) => this.processData(data))
			.catch((error) => Log.error(this.identifier, "Error:", error));
	},
	getParams: function () {
		let params = "?";
		if (this.config.channelIds) {
			params += "id=" + this.config.channelIds.join(",");
		}
		params += "&part=snippet,statistics";
		params += "&fields=items/statistics,items/snippet";
		params += "&key=" + this.config.apiKey;
		return params;
	},
	processData: function (data) {
		try {
			// Handle known errors
			// see https://developers.google.com/youtube/v3/docs/errors#channels_youtube.channels.list
			if (data.error) {
				throw data.error.message;
			}
			// Process data
			this.channelsList = data.items;
			if (this.config.enableRotate) {
				this.rotateData();
				this.scheduleRotate();
			} else {
				this.updateDom(this.config.animationSpeed);
			}
		} catch (error) {
			throw "Unable to process data: " + error;
		}
	},
	paginateData: function (items, page = 1, perPage = 10) {
		const offset = perPage * (page - 1);
		const totalPages = Math.ceil(items.length / perPage);
		const paginatedItems = items.slice(offset, perPage * page);
		return {
			previousPage: page - 1 ? page - 1 : totalPages,
			nextPage: totalPages > page ? page + 1 : 1,
			total: items.length,
			totalPages: totalPages,
			items: paginatedItems
		};
	},
	rotateData: function () {
		if (this.config.enableRotate && this.channelsList) {
			this.currentPage = this.currentChannelList ? this.currentChannelList.nextPage : 1;
			this.currentChannelList = this.paginateData(this.channelsList, this.currentPage, this.config.maximumChannels);
		}
		this.updateDom(this.config.animationSpeed);
	},
	getOrMakeArray: function (values) {
		if (!Array.isArray(values)) {
			values = !values ? [] : values.split(",").map((value) => value.trim());
		}
		return values;
	},
	toMetric: function (input) {
		// see humanizer.node MetricNumerals
		if (input <= 0) {
			return input.toString();
		}
		const symbol = ["", "K", "M", "G", "T", "P", "E", "Z", "Y"];
		const exponent = Math.floor(Math.log10(Math.abs(input)) / 3);
		if (exponent !== 0) {
			let number = input * Math.pow(1000, -exponent);
			let digits = 3 - Math.round(number).toString().length;
			return number.toFixed(digits) + symbol[exponent];
		}
		return Math.round(input).toString();
	}
});
